extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,154 @@
+        offsetIndex = columnChunkReader.getOffsetIndex();
+        Assert.assertion(offsetIndex != null, "offsetIndex != null");
+        numPages = offsetIndex.getPageCount();
+
+        // noinspection unchecked
+        pages = (WeakReference<PageCache.IntrusivePage<ATTR>>[]) new WeakReference[numPages];
+        columnPageReaders = new ColumnPageReader[numPages];
+
+        isPageSizeFixed = true;
+        final long firstPageSize;
+        if (numPages > 1) {
+            firstPageSize = offsetIndex.getFirstRowIndex(1) - offsetIndex.getFirstRowIndex(0);
+        } else {
+            firstPageSize = numRows();
+        }
+        objectsForSynchronizingPageAccess = new Object[numPages];
+        for (int i = 0; i < numPages; ++i) {
+            objectsForSynchronizingPageAccess[i] = new Object();
+            if (isPageSizeFixed && i > 0
+                    && offsetIndex.getFirstRowIndex(i) - offsetIndex.getFirstRowIndex(i - 1) != firstPageSize) {
+                isPageSizeFixed = false;
+            }
+        }
+        if (isPageSizeFixed) {
+            fixedPageSize = firstPageSize;
+        } else {
+            fixedPageSize = -1;
+        }
+    }
+
+    /**
+     * Binary search in offset index to find the page number that contains the row. Logic duplicated from
+     * {@link Arrays#binarySearch(long[], long)} to use the offset index.
+     */
+    private static int findPageNumUsingOffsetIndex(final OffsetIndex offsetIndex, final long row) {
+        int low = 0;
+        int high = offsetIndex.getPageCount() - 1;
+
+        while (low <= high) {
+            final int mid = (low + high) >>> 1;
+            final long midVal = offsetIndex.getFirstRowIndex(mid);
+
+            if (midVal < row)
+                low = mid + 1;
+            else if (midVal > row)
+                high = mid - 1;
+            else
+                return mid; // key found
+        }
+        return -(low + 1); // key not found.
+    }
+
+    private ChunkPage<ATTR> getPage(final int pageNum) {
+        if (pageNum < 0 || pageNum >= numPages) {
+            throw new IllegalArgumentException("pageNum " + pageNum + " is out of range [0, " + numPages + ")");
+        }
+        final PageCache.IntrusivePage<ATTR> page;
+        WeakReference<PageCache.IntrusivePage<ATTR>> pageRef = pages[pageNum];
+        if (pageRef == null || pageRef.get() == null) {
+            synchronized (objectsForSynchronizingPageAccess[pageNum]) {
+                // Make sure no one materialized this page as we waited for the lock
+                pageRef = pages[pageNum];
+                if (pageRef == null || pageRef.get() == null) {
+                    if (columnPageReaders[pageNum] == null) {
+                        columnPageReaders[pageNum] = columnPageReaderIterator.getPageReader(pageNum);
+                    }
+                    try {
+                        page = new PageCache.IntrusivePage<>(toPage(offsetIndex.getFirstRowIndex(pageNum),
```

Comment:
I do kind of wonder if we shouldn't just allocate readers on demand when we are forced to read pages, or just get rid of readers altogether.

You sort of "need" to keep one in the legacy case with no offset indexes, or at least remember the offset index information you constructed during iteration for future use.

In this case, you get nothing but a lock and some garbage from having a reader object.

Maybe the answer is to change the legacy code to just populate offset index data on demand and use that to do your reads.

Note that reader objects currently hold:
1. 7 references that are shared by all readers for a particular column chunk (wasteful)
2. 4 fields specific to the page

I think we could refactor the "read one page" code to live inside a column chunk reader, passing the 4 "per page fields" down as needed. Or, we could have short-lived parameter objects to hold those fields. We ought not to be caching page headers indefinitely, etc.

Then you could do the `updateAndGet` pattern around the actual reading, maybe.

I'm OK with divorcing the "use offset indexes" PR from the "get rid of long-lived reader objects" PR, but I do think we should be making this change; these kinds of overheads add up rapidly at the scale we see in enterprise use-cases.

Reply:
My latest patch makes this change where we created readers only at the time of reading the page and forget them after that. So every time we read the page, we will create a new reader, which is a very small operation.